### PR TITLE
Use a queue for processing incoming envelopes

### LINKED ISSUE #2021: agora.node.Validator.Validator.receiveEnvelope blocks
It seems that `receiveEnvelope` is blocking, meaning the query will wait until completion (probably of some other network query) before returning to the caller.

```
2021-04-23 15:15:06,141 Info [TimeLog.API] - agora.node.Validator.Validator.receiveEnvelope: Elapsed time: 14 ms, 566 μs, and 8 hnsecs
```

Saw that in #1992 , while all the other calls finished in a sub microsecond time.

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
Fixes #2021

### CODE REVIEW

File: source/agora/consensus/protocol/Nominator.d [hunk 1]
```diff
@@ -609,18 +656,46 @@ extern(D):
...
         // Clean the previous envelopes from the DB
-        this.scp_envelope_store.removeAll();
+        this.proc_envelopes_store.removeAll();
 
         // Store the latest envelopes
         foreach (const ref env; envelopes)
-            this.scp_envelope_store.add(env);
+            this.proc_envelopes_store.add(env);
+
+        // Clean the previously queued envelopes from the DDB
```
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d]: DDB -> DB ?

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -437,30 +467,46 @@ extern(D):
...
             envelope = the SCP envelope
 
-        Returns:
-            true if the SCP protocol accepted this envelope
-
     ***************************************************************************/
 
     public void receiveEnvelope (in SCPEnvelope envelope) @trusted
+    {
+        auto dupe = envelope.serializeFull.deserializeFull!SCPEnvelope;
```
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d]: Is "dupe" meaning duplicate? If so maybe copy is better.

File: source/agora/consensus/protocol/Nominator.d [hunk 3]
```diff
@@ -176,9 +186,29 @@ extern(D):
         this.taskman = taskman;
         this.ledger = ledger;
         this.enroll_man = enroll_man;
-        this.scp_envelope_store = this.makeSCPEnvelopeStore(data_dir);
+        this.proc_envelopes_store = this.makeProcessedSCPEnvelopeStore(data_dir);
+        this.queued_envelopes_store = this.makeQueuedSCPEnvelopeStore(data_dir);
         this.restoreSCPState();
         this.nomination_interval = nomination_interval;
+        this.envelope_timer = this.taskman.setTimer(10.msecs,
```
Comment (hewison-chris) [inline source/agora/consensus/protocol/Nominator.d:193]: Perhaps envelope_timer should be a config param for the node.

**Review by hewison-chris — APPROVED** [review]
LGTM just a few comments.

**Comment by Geod24** [comment]
Rebased as https://github.com/bosagora/agora/pull/2166

**Pull request closed without merging**